noir-projects/noir-protocol-circuits/crates/public-kernel-lib/src/components/vm_circuit_output_composer.nr

```diff
@@ -0,0 +1,161 @@
+mod propagate_accumulated_data;
+
+use crate::{components::vm_circuit_output_composer::propagate_accumulated_data::propagate_accumulated_data};
+use dep::types::{
+    abis::{
+    accumulated_data::{PublicAccumulatedDataArrayLengths, PublicAccumulatedDataBuilder},
+    combined_constant_data::CombinedConstantData, gas::Gas,
+    kernel_circuit_public_inputs::vm_circuit_public_inputs::VMCircuitPublicInputs,
+    public_call_request::PublicCallRequest, public_circuit_public_inputs::PublicCircuitPublicInputs,
+    public_data_read::PublicDataRead, public_inner_call_request::PublicInnerCallRequest,
+    validation_requests::{
+    PublicValidationRequestArrayLengths,
+    public_validation_requests_builder::PublicValidationRequestsBuilder
+}
+},
+    constants::MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX, traits::is_empty,
+    utils::arrays::array_to_bounded_vec
+};
+
+// TODO(#7124): To be deprecated.
+struct VMCircuitOutputComposer {
+    constants: CombinedConstantData,
+    call_request: PublicCallRequest,
+    public_call_stack: BoundedVec<PublicInnerCallRequest, MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX>,
+    previous_validation_request_array_lengths: PublicValidationRequestArrayLengths,
+    validation_requests: PublicValidationRequestsBuilder,
+    previous_accumulated_data_array_lengths: PublicAccumulatedDataArrayLengths,
+    accumulated_data: PublicAccumulatedDataBuilder,
+    start_side_effect_counter: u32,
+    end_side_effect_counter: u32,
+    start_gas_left: Gas,
+    transaction_fee: Field,
+    reverted: bool,
+}
+
+impl VMCircuitOutputComposer {
+    pub fn new_from_previous_kernel(previous_kernel: VMCircuitPublicInputs) -> Self {
+        let mut public_call_stack = array_to_bounded_vec(previous_kernel.public_call_stack);
+        let _ = public_call_stack.pop();
+        let validation_requests = PublicValidationRequestsBuilder::new(previous_kernel.validation_requests);
+        let accumulated_data = PublicAccumulatedDataBuilder::new(previous_kernel.accumulated_data);
+        VMCircuitOutputComposer {
+            constants: previous_kernel.constants,
+            call_request: previous_kernel.call_request,
+            public_call_stack,
+            previous_validation_request_array_lengths: previous_kernel.previous_validation_request_array_lengths,
+            validation_requests,
+            previous_accumulated_data_array_lengths: previous_kernel.previous_accumulated_data_array_lengths,
+            accumulated_data,
+            start_side_effect_counter: previous_kernel.start_side_effect_counter,
+            end_side_effect_counter: previous_kernel.end_side_effect_counter,
+            start_gas_left: previous_kernel.start_gas_left,
+            transaction_fee: previous_kernel.transaction_fee,
+            reverted: previous_kernel.reverted
+        }
+    }
+
+    pub fn propagate_from_public_call(&mut self, public_call: PublicCircuitPublicInputs) -> Self {
```

Comment:
It's not super clear to me what the role is of `PublicCircuitPublicInputs` versus `VMCircuitPublicInputs` and whether or not both of them will stay long-term

Reply:
Only the `VMCircuitPublicInputs` will stay. I thought about using `PublicCircuitPublicInputs` for both the output of each individual function call and the entire enqueued call. But that means I would have to modify the vm if I change the struct. So I created a new one instead and leave `PublicCircuitPublicInputs` as is 😂